Feature/display list

benjivou opened the pull request.

benjivou committed:
+ Recycler View with name and ID

benjivou committed:
+ add grid layout

benjivou committed:
+ Scrollable
+ ImageView

benjivou committed:
+ Change the list mode

benjivou committed:
+ Change the list mode

antoine-purnelle-lb commented on app/src/main/java/com/example/movieapp/ListAdapter.kt:
```diff
@@ -0,0 +1,52 @@
+    RecyclerView.ViewHolder(inflater.inflate(R.layout.list_item, parent, false)) {
+    private val TAG = "ListAdapter"
+
+    private var mTitleView: TextView? = null
```
I know it's in a lot of tutorials but the Hungarian notation( putting 'm' at the start of the var name) is horrible and you shouldn't use it

benjivou replied:
done, but what do you prefer ?

benjivou replied:
done

antoine-purnelle-lb commented on app/src/main/java/com/example/movieapp/ListAdapter.kt:
```diff
@@ -0,0 +1,52 @@
+}
+
+class MovieViewHolder(inflater: LayoutInflater, parent: ViewGroup) :
+    RecyclerView.ViewHolder(inflater.inflate(R.layout.list_item, parent, false)) {
```
You can use ViewBiding here too ;-)

benjivou replied:
Sorry I don't understand, what should I do ?

benjivou replied:
Done ( need review )

antoine-purnelle-lb commented on app/src/main/java/com/example/movieapp/MainFragment.kt:
```diff
@@ -0,0 +1,80 @@
+
+    private fun loadListOfMovies(movies: MutableLiveData<List<Movie>>) {
+        list_recycler_view.apply {
+            layoutManager = GridLayoutManager(activity, 2)
```
What about on a tablet? 2 colums won't look good. Search for a way to have a different number of columns in a RV depending on the size of the screen

benjivou replied:
Done

alba-benoit commented on app/src/main/java/com/example/movieapp/ListAdapter.kt:
```diff
@@ -0,0 +1,52 @@
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieViewHolder {
+        val inflater = LayoutInflater.from(parent.context)
+        return MovieViewHolder(inflater, parent)
```
Could be in a single line

alba-benoit commented on app/src/main/java/com/example/movieapp/ListAdapter.kt:
```diff
@@ -0,0 +1,52 @@
+    private var mImage: ImageButton? = null
+
+    init {
+        mTitleView = itemView.findViewById(R.id.list_title)
```
Why did you not use viewBinding ?

benjivou replied:
I already modified this but I don't know why you have the old version on GitHub

antoine-purnelle-lb commented on app/src/main/java/com/example/movieapp/ListAdapter.kt:
```diff
@@ -0,0 +1,52 @@
+        mTitleView = itemView.findViewById(R.id.list_title)
+        mPopularity = itemView.findViewById(R.id.list_popularity)
+        mRate = itemView.findViewById(R.id.list_rate)
+        mImage = itemView.image
```
That line won't compile

benjivou replied:
done

alba-benoit commented on app/src/main/java/com/example/movieapp/ListAdapter.kt:
```diff
@@ -0,0 +1,52 @@
+        mTitleView?.text = movie.title
+        mPopularity?.text = "popularity : ${ movie.popularity}"
+        mRate?.text = "rate  : ${ movie.voteAverage}"
+        Picasso.get().load("https://image.tmdb.org/t/p/w185${movie.posterPath}").into(mImage);
```
"https://image.tmdb.org/t/p/w185" should be a const

benjivou replied:
done

antoine-purnelle-lb commented on app/src/main/res/menu/list_action_bar.xml:
```diff
@@ -0,0 +1,14 @@
+
+    <item
+        android:id="@+id/mode_like"
+        android:title="like" />
```
Those title are not very user friendly ;-)

benjivou replied:
done

alba-benoit commented on app/src/main/java/com/example/movieapp/MainActivity.kt:
```diff
@@ -1,24 +1,9 @@
-class MainActivity : AppCompatActivity() {
 
-    private lateinit var viewModel: MainViewModel
+    override fun createFragment() = MainFragment.newInstance()
```
It would be nice to use Navigation component

benjivou replied:
done

antoine-purnelle-lb commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,3 +1,5 @@
 <resources>
     <string name="app_name">MovieApp</string>
+    <!-- TODO: Remove or change this placeholder text -->
```
Check you TODOs before pushing :-)

benjivou replied:
done

antoine-purnelle-lb commented on app/src/main/res/layout/list_item.xml:
```diff
@@ -0,0 +1,36 @@
+    android:layout_gravity="center_horizontal"
+    android:orientation="vertical"
+    android:padding="16dp">
+    <ImageButton
```
Why an Image**Button**?
A simple ImageView would do the trick

benjivou replied:
done

alba-benoit commented on app/src/main/java/com/example/movieapp/MainFragment.kt:
```diff
@@ -0,0 +1,80 @@
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        retainInstance = true
```
why ?

benjivou replied:
done

antoine-purnelle-lb commented on app/src/main/java/com/example/movieapp/viewModel/MainViewModel.kt:
```diff
@@ -28,24 +28,28 @@ class MainViewModel : ViewModel() {
+    fun getList(typeDisplay: TypeDisplay): MutableLiveData<List<Movie>> {
+        this.typeDisplay = typeDisplay
+        return when (typeDisplay) {
             com.example.movieapp.viewModel.TypeDisplay.POPULAR -> {
```
No need for com.example.movieapp.viewModel.

benjivou replied:
done

antoine-purnelle-lb commented on app/src/main/java/com/example/movieapp/viewModel/MainViewModel.kt:
```diff
@@ -28,24 +28,28 @@ class MainViewModel : ViewModel() {
+                if (rated == null) rated = internetCall(TypeList.HIGHEST_RATE)
                 rated
             }
             com.example.movieapp.viewModel.TypeDisplay.LIKED -> TODO()
```
Good to use the TODO() method but even better if you give it a message so explain why it will crash ;-)

benjivou replied:
thanks done

antoine-purnelle-lb commented on app/src/main/java/com/example/movieapp/viewModel/MainViewModel.kt:
```diff
@@ -28,24 +28,28 @@ class MainViewModel : ViewModel() {
+        return when (typeDisplay) {
             com.example.movieapp.viewModel.TypeDisplay.POPULAR -> {
-                if (popular== null ) popular = internetCall(TypeList.POPULAR)
+                if (popular == null) popular = internetCall(TypeList.POPULAR)
```
It's good to reuse what you've already loaded but you have to give the user a way to refresh. Usually, we use a swipe to refresh. Plenty of examples on the  Internet ;-)

benjivou replied:
done

chrisevrard commented on app/src/main/java/com/example/movieapp/MainFragment.kt:
```diff
@@ -0,0 +1,80 @@
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        retainInstance = true
```
Why do you need to retain the instance?

benjivou replied:
done

chrisevrard commented on app/src/main/java/com/example/movieapp/SingleFragmentActivity.kt:
```diff
@@ -0,0 +1,38 @@
+        var fragment = fm.findFragmentById(R.id.fragment_container)
+        viewModel = ViewModelProvider(
+            this,
+            ViewModelProvider.NewInstanceFactory()
```
Is it useful to create a new instance?

benjivou replied:
done

chrisevrard commented on app/src/main/java/com/example/movieapp/viewModel/MainViewModel.kt:
```diff
@@ -28,24 +28,28 @@ class MainViewModel : ViewModel() {
      */
     fun getListCurrent(): MutableLiveData<List<Movie>> {
-        return when (TypeDisplay) {
+        return getList(typeDisplay)
```
always return a LiveData not a MutableLiveData

benjivou replied:
done

chrisevrard commented on app/src/main/java/com/example/movieapp/SingleFragmentActivity.kt:
```diff
@@ -0,0 +1,38 @@
+
+        setContentView(layoutResId)
+
+        val fm = supportFragmentManager
```
fm is not needed

benjivou replied:
ok done

chrisevrard commented on app/src/main/java/com/example/movieapp/SingleFragmentActivity.kt:
```diff
@@ -0,0 +1,38 @@
+import com.example.movieapp.viewModel.MainViewModel
+
+abstract class SingleFragmentActivity : AppCompatActivity() {
+    lateinit var viewModel: MainViewModel
```
Why do you need to have access to MainViewModel in the activity?

benjivou replied:
No need, job done

chrisevrard commented on app/src/main/java/com/example/movieapp/MainFragment.kt:
```diff
@@ -0,0 +1,80 @@
+        list_recycler_view.apply {
+            layoutManager = GridLayoutManager(activity, 2)
+            movies
+                .observe(viewLifecycleOwner, Observer { adapter = ListAdapter(it) })
```
Don't create an adapter each time

benjivou replied:
done

chrisevrard commented on app/src/main/java/com/example/movieapp/ListAdapter.kt:
```diff
@@ -0,0 +1,52 @@
+
+    fun bind(movie: Movie) {
+        mTitleView?.text = movie.title
+        mPopularity?.text = "popularity : ${ movie.popularity}"
```
hardcoded value

benjivou replied:
done

benjivou committed:
+ Refactoring in process

benjivou committed:
+ strings ressouces values for popularity and rate

benjivou committed:
+ center and add string values

benjivou committed:
+ Use Binding

benjivou committed:
+ CamelCase for the list adapter

benjivou committed:
+ view depending of the screensize

benjivou committed:
+ refactoring code

benjivou committed:
+ refactoring codde
+ replace ImageButton by ImageView

benjivou committed:
+ refactoring

benjivou committed:
+ refactoring String values

benjivou committed:
+ refactoring front

benjivou committed:
+ update every time the list of elements

benjivou committed:
+ reuse the gridlayout

benjivou committed:
+ Refactoring code to make it more readeable

benjivou committed:
+ refactoring res to res+placeholders

alba-benoit commented on app/src/main/java/com/example/movieapp/ListAdapter.kt:
```diff
@@ -0,0 +1,58 @@
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MovieViewHolder {
+        val inflater = LayoutInflater.from(parent.context)
+        return MovieViewHolder(inflater.inflate(R.layout.list_item, parent, false))
```
return MovieViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.list_item, parent, false))

alba-benoit commented on app/src/main/java/com/example/movieapp/ListAdapter.kt:
```diff
@@ -0,0 +1,58 @@
+
+    override fun onBindViewHolder(holder: MovieViewHolder, position: Int) {
+        val movie: Movie = list[position]
+        holder.bind(movie)
```
holer.bind(list[position])

alba-benoit commented on app/src/main/java/com/example/movieapp/MainFragment.kt:
```diff
@@ -0,0 +1,102 @@
+            layoutManager =
+
+                // layout manager for tablet( or device above 7 inches screens
+                if (this.let { activity?.let { it1 -> isTablet(it1) } }!!) GridLayoutManager(
```
weird construction

alba-benoit replied:
if (isTablet(requireActivity)) {
} else {
}

alba-benoit commented on app/src/main/java/com/example/movieapp/MainFragment.kt:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    private fun loadListOfMovies(movies: LiveData<List<Movie>>) {
+        listRecyclerView.apply {
```
Not what we expected

alba-benoit commented on app/src/main/java/com/example/movieapp/frontUtil/FrontUtil.kt:
```diff
@@ -0,0 +1,28 @@
+/**
+ * Created by Benjamin Vouillon on 13,July,2020
+ */
+fun isTablet(activity: Activity): Boolean {
```
There is a easiest way: 
https://developer.android.com/training/multiscreen/screensizes
https://developer.android.com/guide/topics/resources/providing-resources

alba-benoit commented on app/src/main/java/com/example/movieapp/viewModel/MainViewModel.kt:
```diff
@@ -18,34 +19,40 @@ enum class TypeDisplay() {
     /**
      * Our 3 lists of Movies
      */
-    var popular: MutableLiveData<List<Movie>>? = null
```
be more precise on you variable names, for ex, popularMoviesLiveData.
Btw we told you that you need only one liveData -> moviesLiveData

benjivou replied:
Cancel

alba-benoit commented on app/src/main/java/com/example/movieapp/ListAdapter.kt:
```diff
@@ -0,0 +1,58 @@
+
+}
+
+class MovieViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
```
create a file for your class

alba-benoit commented on app/src/main/java/com/example/movieapp/ListAdapter.kt:
```diff
@@ -0,0 +1,58 @@
+            listPopularity.text =
+                resources.getString(itemPopularity,movie.popularity.toString())
+
+            listRate.text =
```
Kotlin Android Extensions, no View Binding ?

benjivou replied:
done

alba-benoit commented on app/src/main/java/com/example/movieapp/MainActivity.kt:
```diff
@@ -1,24 +1,31 @@
-
+        setContentView(layoutResId)
+
+        var fragment = supportFragmentManager.findFragmentById(R.id.fragment_container)
```
use navigation component

benjivou replied:
done

benjivou committed:
+ Binding on itemView
+ Code clean up

benjivou committed:
+ isTablet Optimized

benjivou committed:
+ Optimase the gridLayout

benjivou committed:
+ Optimase Usage of the Gridlayout
+ InternetCall Synchronous in the ModelView

benjivou committed:
+ Refactoring

benjivou committed:
+ Add navigation component
+ Add switchmap to do the work
+ remove useless functions
+ code cleanup
+ refactoring of value

benjivou committed:
+ change internet callback to livedata

benjivou committed:
+ refactoring

alba-benoit commented on app/src/main/java/com/example/movieapp/MainFragment.kt:
```diff
@@ -0,0 +1,71 @@
+
+        // setup the RecyclerView
+        listRecyclerView.apply {
+
```
please avoid unnecessary lines break
